src/utils/messaging-outbox/src/executors/outbox_consumption_iteration_planner.rs

```diff
@@ -0,0 +1,295 @@
+
+        // producer A ->
+        //   consumer X -> message 15
+        //   consumer Y -> message 14
+        // producer B ->
+        //   consumer X -> message 12
+        let consumption_boundaries_by_producer = self
+            .select_latest_consumption_boundaries_by_producer()
+            .await?;
+
+        // producer A: >14
+        // producer B: >12
+        let unconsumed_states_by_producer = self.compute_unconsumed_state_by_producer(
+            latest_message_ids_by_producer,
+            consumption_boundaries_by_producer,
+        );
+        if unconsumed_states_by_producer.is_empty() {
+            return Ok(HashMap::new());
+        }
+
+        // Load portion of unprocessed messages, as specified in filters
+        //   producer A: >14
+        //   producer B: >12
+        let unprocessed_messages_by_producer = self
+            .load_unprocessed_messages_by_producer(&unconsumed_states_by_producer)
+            .await?;
+
+        // Compose consumption tasks
+        Ok(self.compose_producer_consumption_tasks(
+            unprocessed_messages_by_producer,
+            unconsumed_states_by_producer,
+        ))
+    }
+
+    async fn select_latest_message_ids_by_producer(
+        &self,
+    ) -> Result<HashMap<String, OutboxMessageID>, InternalError> {
+        // Extract latest (producer, max message id) relation
+        let latest_message_ids_by_producer = self
+            .outbox_message_repository
+            .get_latest_message_ids_by_producer()
+            .await?;
+
+        // Convert into map
+        Ok(latest_message_ids_by_producer
+            .into_iter()
+            .collect::<HashMap<_, _>>())
+    }
+
+    async fn select_latest_consumption_boundaries_by_producer(
+        &self,
+    ) -> Result<HashMap<String, HashMap<String, OutboxMessageID>>, InternalError> {
+        // Extract consumption boundaries for all routes
+        let all_boundaries = async move {
+            let consumptions_stream = self
+                .outbox_message_consumption_repository
+                .list_consumption_boundaries();
+
+            use futures::TryStreamExt;
+            consumptions_stream.try_collect::<Vec<_>>().await
+        }
+        .await?;
+
+        // Organize by producer->consumer hierarchically
+        let mut boundaries_by_producer = HashMap::new();
+        for boundary in all_boundaries {
+            boundaries_by_producer
+                .entry(boundary.producer_name)
+                .and_modify(|by_consumer: &mut HashMap<String, OutboxMessageID>| {
+                    by_consumer.insert(
+                        boundary.consumer_name.clone(),
+                        boundary.last_consumed_message_id,
+                    );
+                })
+                .or_insert_with(|| {
+                    HashMap::from([(boundary.consumer_name, boundary.last_consumed_message_id)])
+                });
+        }
```

Comment:
So you basically have a list of tuples `(P, C, ID)` and you are doing a `group by P`. I think [`itertools::chunk_by`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.chunk_by) combined with the ability to `.collect()` an iterator of tuples into `HashMap` could reduce the amount of code here, something like:
```
let boundaries: Vec<_> = stream.collect();
boundaries.sort_by(|b| b.producer);
boundaries
.into_iter()
.chunk_by(|b| &b.producer)
.map(|(producer, bounds)| (
  producer, 
  bounds.map(|b| (b.consumer, b.last_id)).collect()
))

Reply:
The code certainly got Rustier, but the resulting "simplicity" is not convincing :-)

Reply:
Before I had to clone the consumer names, but now `chunk_by` forces to clone the producer name. It's probably irrelevant, we are talking about <10 records for now.